File: 08 robots - gbm model/python/gbm.py
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "helpers", "python"))
import tse_helpers as H

import numpy as np
import xgboost as xgb

tse = H.tse


def main():
    ticks = H.load_bidask("btc_bidask.csv")

    # The gradient boosted model is trained here and then lives inside the data processor:
    # the engine never inspects it, it only reads the number the processor pushes.
    model = train_model()

    account = H.account("BTC/USDT", tse.StorageRegime.Mem, tse.LogLevel.Off)
    market = account.create_market("MD", tse.MdType.Bidask)
    account.create_simulator("Sim", H.simulator_options())
    account.add_contract("BTC", 1, tse.Instrument.Future, tse.Underlying.Crypto, tse.Venue.Undefined, 10000)
    account.add_input_bidask("GBM", 1, tse.Duration.Minutes, use_gbm(model), market, ["BTC"])
    # A positive forecast means the model expects the price to rise: the robot stays long
    # while the forecast is positive and closes the position once it turns negative.
    account.add_pattern_threshold("ToLong", tse.Duration.Minutes, ["GBM"], tse.Cmp.Ge, 0.0)
    account.add_pattern_threshold("ToShort", tse.Duration.Minutes, ["GBM"], tse.Cmp.Lt, 0.0)
    account.add_rule_market("Entry", tse.RuleType.Entry, H.entry_params(1.0), "ToLong", "BTC")
    account.add_rule_market("Exit", tse.RuleType.Exit, H.exit_params(), "ToShort", "BTC")
    account.add_robot("Strat", ["Entry", "Exit"])
    account.start("Strat")

    for tick in ticks:
        market.push_bidask_by_name("BTC", tick)

    summary = account.get_summary()
    account.close()

    trades = summary.totalNumberOfTrades
    print("netProfit={:.4f} trades={}".format(summary.totalNetProfit, trades))
    return 0


def train_model():
    index = np.arange(256)
    mid = 44000.0 + (index % 40) * 50.0
    spread = 4.0 + (index % 15)
    imbalance = -1.0 + 2.0 * (index % 21) / 20.0
    features = np.column_stack([mid, spread, imbalance]).astype(np.float32)
    labels = imbalance
    dtrain = xgb.DMatrix(features, label=labels)
    params = {"objective": "reg:squarederror", "max_depth": 3, "eta": 0.3}
    return xgb.train(params, dtrain, num_boost_round=20)


def use_gbm(model):
    def processor(storage, contract_id, tick):
        # Book imbalance carries the signal the model was trained on: more volume resting
        # on the bid than on the ask is buying pressure.
        mid = (tick.bid + tick.ask) / 2.0
        spread = tick.ask - tick.bid
        denom = tick.bidVolume + tick.askVolume
        imbalance = (tick.bidVolume - tick.askVolume) / denom if denom != 0.0 else 0.0
        features = np.array([[mid, spread, imbalance]], dtype=np.float32)
        prediction = float(model.predict(xgb.DMatrix(features))[0])
        storage.push(tick.tsNanoseconds, prediction)
        return storage.size() >= 1

    return processor


if __name__ == "__main__":
    sys.exit(main())

File: 08 robots - gbm model/cpp/gbm.cpp
#include "tse_helpers.hpp"

#include <xgboost/c_api.h>

#include <cmath>
#include <cstdio>
#include <memory>
#include <string>
#include <vector>

namespace {

	BoosterHandle trainModel();
	double predictOne(BoosterHandle booster, double mid, double spread, double imbalance);
	tse::BidAskInputProcessor useGbm(std::shared_ptr<void> model);

}

int main()
{
	std::vector<tse::BidAskTick> const ticks {helpers::loadBidask("btc_bidask.csv")};
	tse::setLogLevel(tse::LogLevel::none);

	// The gradient boosted model is trained here and then lives inside the data processor:
	// the engine never inspects it, it only reads the number the processor pushes.
	std::shared_ptr<void> const model {trainModel(), [](void* handle) { XGBoosterFree(handle); }};

	tse::Account account {"BTC/USDT", tse::StorageRegime::mem};
	tse::Market const market {account.createMarket("MD", tse::MdType::bidask)};
	account.createSimulator("Sim", helpers::simulatorConfig());
	account.addContract("BTC", 1, tse::Instrument::future, tse::Underlying::crypto, tse::Venue::undefined, 10000);
	account.addInputBidAsk("GBM", 1, tse::Duration::minutes, useGbm(model), market, {"BTC"});
	// A positive forecast means the model expects the price to rise: the robot stays long
	// while the forecast is positive and closes the position once it turns negative.
	account.addPatternThreshold("ToLong", tse::Duration::minutes, {"GBM"}, tse::Cmp::ge, 0.0);
	account.addPatternThreshold("ToShort", tse::Duration::minutes, {"GBM"}, tse::Cmp::lt, 0.0);
	account.addRuleMarket("Entry", tse::RuleType::entry, helpers::entryParams(1.0), "ToLong", "BTC");
	account.addRuleMarket("Exit", tse::RuleType::exit, helpers::exitParams(), "ToShort", "BTC");
	account.addRobot("Strat", {"Entry", "Exit"});
	account.start("Strat");

	for (tse::BidAskTick const& tick : ticks) {
		market.pushBidAsk("BTC", tick);
	}

	tse::Summary const summary {account.getSummary()};
	long long const trades {static_cast<long long>(summary.totalNumberOfTrades)};
	std::printf("netProfit=%.4f trades=%lld\n", summary.totalNetProfit, trades);
	return 0;
}

namespace {

	BoosterHandle trainModel()
	{
		std::vector<float>
			features(256 * 3, 0.0f),
			labels(256, 0.0f);
		for (std::size_t i {0}; i < 256; ++i) {
			float const
				mid       {44000.0f + static_cast<float>(i % 40) * 50.0f},
				spread    {4.0f + static_cast<float>(i % 15)},
				imbalance {-1.0f + 2.0f * static_cast<float>(i % 21) / 20.0f};
			features[i * 3 + 0] = mid;
			features[i * 3 + 1] = spread;
			features[i * 3 + 2] = imbalance;
			labels[i] = imbalance;
		}
		DMatrixHandle dtrain {nullptr};
		XGDMatrixCreateFromMat(features.data(), 256, 3, std::nanf(""), &dtrain);
		XGDMatrixSetFloatInfo(dtrain, "label", labels.data(), 256);

		BoosterHandle booster {nullptr};
		XGBoosterCreate(&dtrain, 1, &booster);
		XGBoosterSetParam(booster, "objective", "reg:squarederror");
		XGBoosterSetParam(booster, "max_depth", "3");
		XGBoosterSetParam(booster, "eta", "0.3");
		for (int round {0}; round < 20; ++round) {
			XGBoosterUpdateOneIter(booster, round, dtrain);
		}
		XGDMatrixFree(dtrain);
		return booster;
	}

	double predictOne
	(
		BoosterHandle booster,
		double mid,
		double spread,
		double imbalance
	)
	{
		float const features[3] {static_cast<float>(mid), static_cast<float>(spread), static_cast<float>(imbalance)};
		DMatrixHandle dmatrix {nullptr};
		XGDMatrixCreateFromMat(features, 1, 3, std::nanf(""), &dmatrix);
		bst_ulong outLen {0};
		float const* outResult {nullptr};
		XGBoosterPredict(booster, dmatrix, 0, 0, 0, &outLen, &outResult);
		double const prediction {outLen > 0 ? static_cast<double>(outResult[0]) : 0.0};
		XGDMatrixFree(dmatrix);
		return prediction;
	}

	tse::BidAskInputProcessor useGbm(std::shared_ptr<void> model)
	{
		return [model]
		(
			tse::Storage const& storage,
			std::string const&,
			tse::BidAskTick const& tick
		) -> bool
		{
			// Book imbalance carries the signal the model was trained on: more volume resting
			// on the bid than on the ask is buying pressure.
			double const
				mid {(tick.bid + tick.ask) / 2.0},
				spread {tick.ask - tick.bid},
				denom {tick.bidVolume + tick.askVolume},
				imbalance {denom != 0.0 ? (tick.bidVolume - tick.askVolume) / denom : 0.0};
			storage.push(tick.tsNanoseconds, predictOne(model.get(), mid, spread, imbalance));
			return storage.size() >= 1;
		};
	}

}
